lib/src/main/java/graphql/nadel/engine/blueprint/NadelExecutionBlueprintFactory.kt

```diff
@@ -771,44 +771,68 @@ private class SharedTypesAnalysis(
         overallParentType: AnyImplementingTypeDefinition,
         underlyingParentType: GraphQLFieldsContainer,
     ): List<NadelTypeRenameInstruction> {
-        val overallOutputTypeName = overallField.type.unwrapAll().name
-
         val underlyingField = getUnderlyingField(overallField, overallParentType, underlyingParentType)
             ?: return emptyList()
 
-        val renameInstruction = if (overallOutputTypeName !in serviceDefinedTypes) {
+        val overallOutputTypeName = overallField.type.unwrapAll().name
+        val underlyingOutputTypeName = underlyingField.type.unwrapAll().name
+        val outputTypeRenameInstruction = getTypeRenameInstructionOrNull(
+            overallTypeName = overallOutputTypeName,
+            underlyingTypeName = underlyingOutputTypeName,
+            serviceDefinedTypes = serviceDefinedTypes,
+            service = service,
+        )
+
+        val argumentTypeRenameInstructions = overallField.inputValueDefinitions
```

Comment:
so we never considered input type renames here?

Reply:
Not shared ones. This code is looking for references to types used by the service, but NOT defined by the service itself.

As we discussed elsewhere, it should really reuse the validation code that uses a schema traverser to pick up all referenced types.